Add option to disable MQTT Alarm Control Panel supported features

paravoid opened the pull request:
## Proposed change

The MQTT Alarm Control Panel currrently enables all features (arm home, arm away, arm night, arm vacation, arm custom bypass) unconditionally. This clutters the interface and can even be potentially dangerous, by enabling modes that the remote alarm may not support.

Make all the features conditional, similar to how other MQTT components (e.g. cover, light etc.) handle this, as well as how other Alarm Panels (e.g. template) do as well.

Keep the default payloads as configured, to **maintain backwards compatibility**. In other words, all modes are (still) enabled by default, but can now be disabled by setting the new `supported_features` option to the list of features supported.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

_(Arguably a bugfix)_

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes # (none reported)
- This PR is related to issue: (none reported)
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/28599

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

~~If the code communicates with devices, web services, or third-party tools:~~

- [ ] ~~The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.~~
- [ ] ~~New or updated dependencies have been added to `requirements_all.txt`. 
      Updated by running `python3 -m script.gen_requirements_all`.~~
- [ ] ~~For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.~~
- [ ] ~~Untested files have been added to `.coveragerc`.~~

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jbouwh reviewed (changes requested):
@paravoid I like the Idea to be able to limit the features, but I am not sure using `None` for the payload options is the best option to do this,
May be a new option `supported_features` (CONF_SUPPORTED_FEATURES)  (there is already an abbreviation) suits better for this goal.
It should configure a list with default value `["arm_away", "arm_custom_bypass","arm_home","arm_night","arm_vacation","trigger"]`
With a mapping of the strings to the feature flags you can calculate the supported features.

paravoid commented:
@jbouwh thank you for the review! I see how many PRs are open, so really appreciate the effort and promptness here!

I think I understand what you're saying and, honestly, I agree with you: empty payloads don't feel very intuitive, and something like your proposed solution feels more explicit, obvious and future-proof.

The reason I went with the current approach, however, is that upon researching this, I saw that many (most? all?) other MQTT components implement this this way. I didn't look at all of them, but I started alphabetically and looked at (at least) Climate, Cover, Light and Fan.

For example Climate has:
```python
        support = ClimateEntityFeature(0)
        if (self._topic[CONF_TEMP_STATE_TOPIC] is not None) or (
            self._topic[CONF_TEMP_COMMAND_TOPIC] is not None
        ):
            support |= ClimateEntityFeature.TARGET_TEMPERATURE
        [...]
```

Cover:
```python
        supported_features = CoverEntityFeature(0)
        if self._config.get(CONF_COMMAND_TOPIC) is not None:
            if self._config.get(CONF_PAYLOAD_OPEN) is not None:
                supported_features |= CoverEntityFeature.OPEN
            if self._config.get(CONF_PAYLOAD_CLOSE) is not None:
                supported_features |= CoverEntityFeature.CLOSE
            [...]
```

...and so on and so forth.

Of course two wrongs don't make a right, but thinking about intuitiveness, I figured that consistency is probably key and the most intuitive way is the one that users are used to, from other MQTT components they may have configured.

Given this information, what do you think? Do you think I should deviate from what seems to be the existing practice in this case?

frenck commented:
One thing to note: supported feature are not allowed to change during runtime of Home Assistant (with the exception of media players).

jbouwh commented:
> One thing to note: supported feature are not allowed to change during runtime of Home Assistant (with the exception of media players).

Right, like other conf features this should be a config option that is set during setup of the entity.

jbouwh commented:
> @jbouwh thank you for the review! I see how many PRs are open, so really appreciate the effort and promptness here!
> 
> I think I understand what you're saying and, honestly, I agree with you: empty payloads don't feel very intuitive, and something like your proposed solution feels more explicit, obvious and future-proof.
> 
> The reason I went with the current approach, however, is that upon researching this, I saw that many (most? all?) other MQTT components implement this this way. I didn't look at all of them, but I started alphabetically and looked at (at least) Climate, Cover, Light and Fan.
> 
> For example Climate has:
> 
> ```python
>         support = ClimateEntityFeature(0)
>         if (self._topic[CONF_TEMP_STATE_TOPIC] is not None) or (
>             self._topic[CONF_TEMP_COMMAND_TOPIC] is not None
>         ):
>             support |= ClimateEntityFeature.TARGET_TEMPERATURE
>         [...]
> ```
> 
> Cover:
> 
> ```python
>         supported_features = CoverEntityFeature(0)
>         if self._config.get(CONF_COMMAND_TOPIC) is not None:
>             if self._config.get(CONF_PAYLOAD_OPEN) is not None:
>                 supported_features |= CoverEntityFeature.OPEN
>             if self._config.get(CONF_PAYLOAD_CLOSE) is not None:
>                 supported_features |= CoverEntityFeature.CLOSE
>             [...]
> ```
> 
> ...and so on and so forth.
> 
> Of course two wrongs don't make a right, but thinking about intuitiveness, I figured that consistency is probably key and the most intuitive way is the one that users are used to, from other MQTT components they may have configured.
> 
> Given this information, what do you think? Do you think I should deviate from what seems to be the existing practice in this case?

What you are saying about other platforms applies to features that are set when command topics are configured, but we can't use that here because the tropics are shared.

jbouwh commented:
If you find it difficult to implement the suggestions I did, I'll be happy to help.

paravoid commented:
> What you are saying about other platforms applies to features that are set when command topics are configured, but we can't use that here because the tropics are shared.

I did a bit more comprehensive review of the MQTT component, and sadly that's not actually the case consistently.
(Sorry if it sounds like I'm insisting, just trying to get this right.Happy to go with the path of least resistance if you don't think that's a valuable path to be going down. )

Option 1: Separate command topics, features enabled based on the existence of the configured command topics:
* Climate
* Fan
* Update (sort of)
* Light (sort of)

Option 2: Shared command topic, features enabled based on the payloads configured, with None = feature disabled:
* Cover (see excerpt quoted above)
* Lock

Option 3: Boolean configuration options to enable/disable features
* Siren: `support_duration`, `support_volume_set`, both True/False

Option 4: Separate list of strings configuration for supported features:
* Humidifier: `modes` (also allows arbitrary values for custom modes)

In this PR I've basically implemented option (2) and what you're requested is, I believe, a variation of option (4). Am I getting it right?

> If you find it difficult to implement the suggestions I did, I'll be happy to help.

Thanks for the offer! You're already helping :) I think I've got it, it may even be simpler than what I've implemented here?

jbouwh commented:
> Thanks for the offer! You're already helping :) I think I've got it, it may even be simpler than what I've implemented here?

It probably will be.

paravoid committed:
Make MQTT Alarm Control Panel features conditional

The MQTT Alarm Control Panel currently enables all features (arm home,
arm away, arm night, arm vacation, arm custom bypass) unconditionally.
This clutters the interface and can even be potentially dangerous, by
enabling modes that the remote alarm may not support.

Make all the features conditional, by adding a new "supported_features"
configuration option, comprising a list of the supported features as
options. Feature enablement seems inconsistent across the MQTT
component; this implementation is most alike to the Humidifier modes
option, but using a generic "supported_features" name that other
implementations may reuse in the future.

The default value of this new setting remains to be all features, which
while it may be overly expansive, is necessary to maintain backwards
compatibility.

paravoid commented:
OK, from the terseness of the response, I take it you'd rather go with what you originally suggested and for me to not continue down the path of consistency :) Can do!

PR updated - hopefully this matches what you envisioned. Let me know what you think, and I'll submit the documentation PR as well for the new option as well :)

jbouwh commented on tests/components/mqtt/test_alarm_control_panel.py:
```diff
@@ -223,6 +233,53 @@ async def test_ignore_update_state_if_unknown_via_state_topic(
+                DEFAULT_CONFIG,
+                ({"supported_features": []},),
+            ),
+            0,
```
```suggestion
            AlarmControlPanelEntityFeature(0),
```

Do we want to allow no supported feature is set?

paravoid replied:
Ack on the `0` -> `AlarmControlPanelEntityFeature(0)` change, makes sense!

> Do we want to allow no supported feature is set?

I don't see why not. That's the default in the base component, as well as, as an example, in the template alarm panel integration.

If anything I'd argue that "disarm" should also be a feature that can be disabled, to also be able to support read-only alarm integrations - such as the case of systems with an event bus or programmable outputs. But that's not supported in the core integration, so kinda besides the point :)

jbouwh reviewed (commented):
Looks already very promising. I'd like to do some more testing. Can you prepare a docs PR too?

jbouwh commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -217,14 +230,10 @@ def state(self) -> str | None:
-            | AlarmControlPanelEntityFeature.ARM_AWAY
-            | AlarmControlPanelEntityFeature.ARM_NIGHT
-            | AlarmControlPanelEntityFeature.ARM_VACATION
-            | AlarmControlPanelEntityFeature.ARM_CUSTOM_BYPASS
-            | AlarmControlPanelEntityFeature.TRIGGER
-        )
+        supported_features = AlarmControlPanelEntityFeature(0)
+        for feature in self._config[CONF_SUPPORTED_FEATURES]:
+            supported_features |= _SUPPORTED_FEATURES[feature]
+        return supported_features
```
May be we can assign `self._attr_supported_features` in `self._setup_from_config` and remove the property so it is not iterating every time.

paravoid replied:
Oooh makes sense! Will do.

paravoid commented:
Thanks for the review! I think I've addressed all of your comments & also pushed the documentation PR as well (not sure why CI still says "docs-missing").

jbouwh commented:
Please do not overwrite the old commit, it makes it hard to see what you have changed. It is ok if you want to force push them on a rebase. I'll have a closer tomorrow.

jbouwh commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -81,6 +91,9 @@
 
 PLATFORM_SCHEMA_MODERN = MQTT_BASE_SCHEMA.extend(
     {
+        vol.Required(
```
I do not think we should make this option required, because of the default value this is not needed.

paravoid replied:
I debated that myself, wondering about the differences between Optional w/ default vs. Required w/ default. From a user configuration perspective they seemed equivalent.

However, I opted into Required because technically it _is_ actually required for the code to work: we do `self._config[CONF_SUPPORTED_FEATURES]` later on, rather than `self._config.get(CONF_SUPPORTED_FEATURES)` etc. Static analysis like mypy isn't smart enough to pick this up, but it could in the future. (git grepping Required.*default results into 384 matches across the tree, so it's definitely not unheard of either)

jbouwh replied:
Introducing a new Required option is weird as it is absent in all configurations from the user perspective. Because it has a default value, it is not required in the user config or discovery payload. Technically you are right that it does not make a difference.

paravoid replied:
Could you help me understand why it's weird? If you can elaborate on what you fear would happen, or any references, or prior discussions.

My instincts suggest this is better for the reasons I stated, but given I'm a new contributor, my instincts are not fine-tuned for this codebase :) But when I tried looking to get a better feel of how others have approached this, I found lots of callsites doing it this way, even in very popular or core parts of HA, such as for example the Integration helper (CONF_ROUND_DIGITS, CONF_METHOD, CONF_UNIT_PREFIX, CONF_UNIT_TIME), Bluetooth (CONF_PASSIVE), ZHA, ZWave-js, etc. Would love to hear more to make this a learning experience for me.

jbouwh replied:
It is not clear to the users to make it required while it is not. It should be consistent with current configs that it is not needed to add the option. I cannot make it more clear than that.

paravoid replied:
OK, let me try again: my understanding is that `vol.Required` = "_this code requires this attribute to function_", rather than "_the user needs to specify this option_". In fact given we provide a default here, a user-provided value is explicitly NOT required here. Is my understanding incorrect?

As for consistency, I already gave several examples in the codebase of options that are not explicitly needed, that are configured in the same way. So I don't understand the consistency argument. Could you help explain how this option is different than e.g. the Integration configuration options I mentioned above?

jbouwh replied:
We only use `Required` when there is no default and the schema will fail without it, so hat is why it needs to change. The required flag in the documentation needs to be inline with the code. It is hard to explain why an option is marked as `Required` while it is not. That is an unneeded restriction. Hope it is clear enough.

paravoid replied:
As I mentioned, there are multiple counter-examples in the tree, so the statements you made ("_We only use `Required` when there is no default_") do not seem to match the existing practice, at least not consistently. A mere `git grep 'vol\.Required.*default'` is enough to validate this.

All that said, I think we've both spent way more time on this than it deserved, so I'll do as you asked.

Should I also send a PR to https://developers.home-assistant.io/docs/development_validation/ under the "Some things to keep in mind" section that says something like "_Do not ever use `vol.Required` with a default setting; use vol.Optional instead_"? Happy for you to tweak the wording as you please, and even happier if you'd like to submit that PR yourself. Just looking out for everyone's else time on this :)

jbouwh commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -81,6 +91,9 @@
 
 PLATFORM_SCHEMA_MODERN = MQTT_BASE_SCHEMA.extend(
     {
+        vol.Required(
```
```suggestion
        vol.Optional(
```

jbouwh commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -81,6 +91,9 @@
     {
+        vol.Required(
+            CONF_SUPPORTED_FEATURES, default=list(_SUPPORTED_FEATURES)
+        ): cv.multi_select(_SUPPORTED_FEATURES),
```
The docs PR mentions you can also set a single string. If you want to implement this you should add `cv.ensure_list` in advance to `cv.multi_select()`.

jbouwh commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -139,6 +152,9 @@ class MqttAlarm(MqttEntity, alarm.AlarmControlPanelEntity):
     _default_name = DEFAULT_NAME
     _entity_id_format = alarm.ENTITY_ID_FORMAT
     _attributes_extra_blocked = MQTT_ALARM_ATTRIBUTES_BLOCKED
+    _attr_supported_features: AlarmControlPanelEntityFeature = (
+        AlarmControlPanelEntityFeature(0)
+    )
```
Can you move the initialization to `_setup_from_config` too?

jbouwh replied:
```suggestion
    _attr_supported_features: AlarmControlPanelEntityFeature
```

jbouwh reviewed (changes requested):
Please address requested changes before making the PR ready for review again.

paravoid committed:
Apply suggestions from code review

* Use vol.Optional() instead of vol.Required() for "supported_features".
* Move the initialization of _attr_supported_features to _setup_from_config.

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

jbouwh commented:
It is correct that not all MQTT platform implementations follow the same architecture. Aligning them takes a lot of time when this causes breaking changes. The MQTT `alarm_control_panel` platform will be less frequently used than e.g. `sensor` or `light`. We try to get the support as optimal as possible. Thnx for your patience.

jbouwh reviewed (dismissed):
Nice, thank you for your contribution
@paravoid 👍 

@emontnemery can you do a final review?

jbouwh commented:
@paravoid may be you can update the (one of the) examples in the docs PR to demonstrate the use of the new option.

emontnemery commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -139,6 +152,7 @@ class MqttAlarm(MqttEntity, alarm.AlarmControlPanelEntity):
     _default_name = DEFAULT_NAME
     _entity_id_format = alarm.ENTITY_ID_FORMAT
     _attributes_extra_blocked = MQTT_ALARM_ATTRIBUTES_BLOCKED
+    _attr_supported_features: AlarmControlPanelEntityFeature
```
This already exists in the base class
```suggestion
```

paravoid replied:
Good catch!

emontnemery commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -167,6 +181,10 @@ def _setup_from_config(self, config: ConfigType) -> None:
             config[CONF_COMMAND_TEMPLATE], entity=self
         ).async_render
 
+        self._attr_supported_features = AlarmControlPanelEntityFeature(0)
```
This is already initialized in the base class
```suggestion
```

paravoid committed:
Apply suggestions from emontnemery's code review

* Use vol.In() instead of cv.multi_seelct()
* Remove superfluous _attr_supported_features initializers, already
  present in the base class.

Co-authored-by: Erik Montnemery <erik@montnemery.com>

paravoid committed:
Add invalid config tests for the MQTT Alarm Control Panel

paravoid commented on homeassistant/components/mqtt/alarm_control_panel.py:
```diff
@@ -81,6 +91,9 @@
     {
+        vol.Optional(
+            CONF_SUPPORTED_FEATURES, default=list(_SUPPORTED_FEATURES)
+        ): cv.multi_select(_SUPPORTED_FEATURES),
```
Oh nifty, thanks!

jbouwh commented on tests/components/mqtt/test_alarm_control_panel.py:
```diff
@@ -264,20 +267,56 @@ async def test_ignore_update_state_if_unknown_via_state_topic(
+                DEFAULT_CONFIG,
+                ({"supported_features": ["invalid"]},),
+            ),
+            AlarmControlPanelEntityFeature.ARM_HOME
+            | AlarmControlPanelEntityFeature.ARM_AWAY,
```
You are not using these fro the False case. So you can use `None`

jbouwh commented on tests/components/mqtt/test_alarm_control_panel.py:
```diff
@@ -264,20 +267,56 @@ async def test_ignore_update_state_if_unknown_via_state_topic(
+                DEFAULT_CONFIG,
+                ({"supported_features": "invalid"},),
+            ),
+            AlarmControlPanelEntityFeature.ARM_HOME
+            | AlarmControlPanelEntityFeature.ARM_AWAY,
```
You are not using these fro the False case. So you can assign `None`

jbouwh replied:
Did you miss the other 2 occurrences?

paravoid replied:
Yeah I did... So sorry about that, I got pretty tired at that time of the day and with so many back and forths. Thank you for catching this!

jbouwh commented on tests/components/mqtt/test_alarm_control_panel.py:
```diff
@@ -264,20 +267,56 @@ async def test_ignore_update_state_if_unknown_via_state_topic(
 async def test_supported_features(
     hass: HomeAssistant,
     mqtt_mock_entry: MqttMockHAClientGenerator,
     expected_features: AlarmControlPanelEntityFeature,
```
Add `| None` as type hint.

jbouwh reviewed (changes requested):
A few comments on the added False tests

paravoid committed:
Set expected_features to None in the invalid MQTT Alarm Control Panel tests

jbouwh commented on tests/components/mqtt/test_alarm_control_panel.py:
```diff
@@ -223,6 +233,90 @@ async def test_ignore_update_state_if_unknown_via_state_topic(
+                DEFAULT_CONFIG,
+                ({"supported_features": "invalid"},),
+            ),
+            AlarmControlPanelEntityFeature.ARM_HOME
+            | AlarmControlPanelEntityFeature.ARM_AWAY,
```
```suggestion
            None,
```

paravoid committed:
Add another expected_features=None in the invalid tests

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

paravoid commented:
Shall I squash all these commits myself now, or does that happen automatically during merge time? Same question for rebasing against latest dev.

jbouwh commented:
> Shall I squash all these commits myself now, or does that happen automatically during merge time? Same question for rebasing against latest dev.

Don't worry happens during merging time.

jbouwh reviewed (approved):
LGTM, nice first contribution,
Thnx @paravoid 👍

emontnemery reviewed (approved):
Thanks, @paravoid 👍
